### DISCUSSION TITLE
Searching in facet indices

### DISCUSSION THREAD

**Discussion by francoischalifour** [algolia/instantsearch#4400] [Ideas]
Since InstantSearch.js v4, we introduced the [`index`](https://www.algolia.com/doc/api-reference/widgets/index-widget/js/) widget that allows to target multiple Algolia indices in a single [`instantsearch`](https://www.algolia.com/doc/api-reference/widgets/instantsearch/js/) instance—and **what if we allowed to also target facet indices**?

## Motivation

When building advanced search interfaces, you often need to **search through "hidden" data** at first glance (e.g., hit categories) to allow to **centralize all filters in a single search box**.

Here's an example in a Pokédex application, where I search Pokémon and their abilities at the same time, to allow to refine the search.

![pokedex-facet-index-search](https://user-images.githubusercontent.com/6137112/80607976-72f8da80-8a36-11ea-9a99-32ec07096b60.gif)

I had to use the search client for this. As a user, I would expect to be able to **use a `facetIndex` widget to search** through these values.

## Proposed usage

A facet index could act as an index, where its widgets inherit from its context: a nested search box would target a facet index (through a [SFFV](https://www.algolia.com/doc/api-reference/api-methods/search-for-facet-values/) endpoint), a `hits` widget would display facet hits, etc.

Using the React syntax for convenience:

```jsx
function App() {
  return (
    <InstantSearch>
      <FacetIndex facetName="brand">
        <Configure maxFacetHits={5} />

        {/* Sets the `facetQuery` search parameter when in a facet index */}
        <SearchBox />

        {/* Reads from the `facetHits` response when in a facet index */}
        <Hits />
      </FacetIndex>
    </InstantSearch>
  );
}
```

Am I the only one in need for such a feature?

**Comment by Haroenv** [comment]
How do the other widgets make sense in a "facet index"? Would they just throw an error?

**Reply by francoischalifour** [reply]
I would say so. What would you see?

**Reply by Haroenv** [reply]
in InstantSearch.js I would expect that error to happen in `addWidgets`, it could filter accepted child widgets for FacetIndex begin hits / searchbox, maybe refinementList & menu on the attribute you're in (to be able to refine)

**Reply by francoischalifour** [reply]
If we take things further, we could even statically prevent users from these errors with ESLint rules.
